File: training/train_matte_refiner.mjs
/**
 * Matte refiner trainer (pure Node.js — no PyTorch). Weights: private/models/matte-refiner.json
 */
import { mkdir, readdir, readFile, writeFile } from 'node:fs/promises';
import path from 'node:path';
import sharp from 'sharp';

const ROOT = path.resolve(import.meta.dirname, '..');
const config = JSON.parse(await readFile(path.join(ROOT, 'training/config.json'), 'utf8'));
const LABELS = path.resolve(ROOT, config.labelsDir);
const MODEL_DIR = path.resolve(ROOT, config.modelDir);
const CKPT_DIR = path.resolve(ROOT, config.checkpointDir);
const REPORT_DIR = path.resolve(ROOT, config.reportDir);
const SIZE = 64;
const HIDDEN = 192;
const EPOCHS = Number(config.epochs);
const LR = Number(config.learningRate);

const INPUT_DIM = SIZE * SIZE * 3;
const OUTPUT_DIM = SIZE * SIZE;

function sigmoid(x) {
  if (x >= 0) {
    const z = Math.exp(-x);
    return 1 / (1 + z);
  }
  const z = Math.exp(x);
  return z / (1 + z);
}

function randn(scale) {
  return (Math.random() * 2 - 1) * scale;
}

function initMatrix(rows, cols, scale) {
  const data = new Float32Array(rows * cols);
  for (let i = 0; i < data.length; i += 1) {
    data[i] = randn(scale);
  }
  return data;
}

function initVector(len) {
  return new Float32Array(len);
}

function matVec(M, rows, cols, v) {
  const out = new Float32Array(rows);
  for (let r = 0; r < rows; r += 1) {
    let sum = 0;
    const row = r * cols;
    for (let c = 0; c < cols; c += 1) {
      sum += (M[row + c] ?? 0) * (v[c] ?? 0);
    }
    out[r] = sum;
  }
  return out;
}

class MLP {
  constructor() {
    this.w1 = initMatrix(HIDDEN, INPUT_DIM, Math.sqrt(2 / INPUT_DIM));
    this.b1 = initVector(HIDDEN);
    this.w2 = initMatrix(OUTPUT_DIM, HIDDEN, Math.sqrt(2 / HIDDEN));
    this.b2 = initVector(OUTPUT_DIM);
  }

  forward(x) {
    const z1 = matVec(this.w1, HIDDEN, INPUT_DIM, x);
    for (let i = 0; i < HIDDEN; i += 1) {
      z1[i] = (z1[i] ?? 0) + (this.b1[i] ?? 0);
    }
    const h = new Float32Array(HIDDEN);
    for (let i = 0; i < HIDDEN; i += 1) {
      h[i] = Math.max(0, z1[i] ?? 0);
    }
    const z2 = matVec(this.w2, OUTPUT_DIM, HIDDEN, h);
    for (let i = 0; i < OUTPUT_DIM; i += 1) {
      z2[i] = (z2[i] ?? 0) + (this.b2[i] ?? 0);
    }
    const y = new Float32Array(OUTPUT_DIM);
    for (let i = 0; i < OUTPUT_DIM; i += 1) {
      y[i] = sigmoid(z2[i] ?? 0);
    }
    return { y, h, z1, z2 };
  }

  backward(x, cache, target) {
    const { y, h, z2 } = cache;
    const gradY = new Float32Array(OUTPUT_DIM);
    for (let i = 0; i < OUTPUT_DIM; i += 1) {
      gradY[i] = (y[i] ?? 0) - (target[i] ?? 0);
    }

    const gradZ2 = new Float32Array(OUTPUT_DIM);
    for (let i = 0; i < OUTPUT_DIM; i += 1) {
      const s = y[i] ?? 0;
      gradZ2[i] = (gradY[i] ?? 0) * s * (1 - s);
    }

    for (let i = 0; i < OUTPUT_DIM; i += 1) {
      this.b2[i] -= LR * (gradZ2[i] ?? 0);
      const row = i * HIDDEN;
      for (let j = 0; j < HIDDEN; j += 1) {
        this.w2[row + j] -= LR * (gradZ2[i] ?? 0) * (h[j] ?? 0);
      }
    }

    const gradH = new Float32Array(HIDDEN);
    for (let j = 0; j < HIDDEN; j += 1) {
      let sum = 0;
      for (let i = 0; i < OUTPUT_DIM; i += 1) {
        sum += (gradZ2[i] ?? 0) * (this.w2[i * HIDDEN + j] ?? 0);
      }
      gradH[j] = (h[j] ?? 0) > 0 ? sum : 0;
    }

    for (let j = 0; j < HIDDEN; j += 1) {
      this.b1[j] -= LR * (gradH[j] ?? 0);
      const row = j * INPUT_DIM;
      for (let k = 0; k < INPUT_DIM; k += 1) {
        this.w1[row + k] -= LR * (gradH[j] ?? 0) * (x[k] ?? 0);
      }
    }
  }

  toJSON() {
    return {
      version: 1,
      architecture: 'mlp',
      trainSize: SIZE,
      hidden: HIDDEN,
      w1: [...this.w1],
      b1: [...this.b1],
      w2: [...this.w2],
      b2: [...this.b2],
    };
  }
}

async function loadSample(originalPath, maskPath) {
  const rgbBuf = await sharp(originalPath).resize(SIZE, SIZE).removeAlpha().raw().toBuffer();
  const maskBuf = await sharp(maskPath).resize(SIZE, SIZE).grayscale().raw().toBuffer();
  const input = new Float32Array(INPUT_DIM);
  for (let i = 0; i < SIZE * SIZE; i += 1) {
    input[i] = (rgbBuf[i * 3] ?? 0) / 255;
    input[SIZE * SIZE + i] = (rgbBuf[i * 3 + 1] ?? 0) / 255;
    input[2 * SIZE * SIZE + i] = (rgbBuf[i * 3 + 2] ?? 0) / 255;
  }
  const target = new Float32Array(OUTPUT_DIM);
  for (let i = 0; i < OUTPUT_DIM; i += 1) {
    target[i] = (maskBuf[i] ?? 0) / 255;
  }
  return { input, target };
}

async function listSamples() {
  const split = path.resolve(ROOT, config.privateRoot, 'splits/train.txt');
  const ids = (await readFile(split, 'utf8')).split(/\r?\n/).filter(Boolean);
  const samples = [];
  for (const id of ids) {
    const dir = path.join(LABELS, id);
    const files = await readdir(dir);
    const origName = files.find((n) => /^original\./i.test(n));
    if (!origName) continue;
    samples.push({ id, originalPath: path.join(dir, origName), finalPath: path.join(dir, 'final.png') });
  }
  return samples;
}

function mse(pred, target) {
  let sum = 0;
  for (let i = 0; i < pred.length; i += 1) {
    const d = (pred[i] ?? 0) - (target[i] ?? 0);
    sum += d * d;
  }
  return sum / pred.length;
}

await mkdir(MODEL_DIR, { recursive: true });
await mkdir(CKPT_DIR, { recursive: true });
await mkdir(REPORT_DIR, { recursive: true });

const sampleMeta = await listSamples();
if (sampleMeta.length === 0) {
  console.error('No samples. Run: npm run train:bootstrap');
  process.exit(1);
}

const loaded = await Promise.all(
  sampleMeta.map(async (s) => ({
    id: s.id,
    ...(await loadSample(s.originalPath, s.finalPath)),
  })),
);

const model = new MLP();
let best = Number.POSITIVE_INFINITY;
const history = [];

for (let epoch = 1; epoch <= EPOCHS; epoch += 1) {
  let total = 0;
  for (const sample of loaded) {
    const cache = model.forward(sample.input);
    total += mse(cache.y, sample.target);
    model.backward(sample.input, cache, sample.target);
  }
  const avg = total / loaded.length;
  history.push({ epoch, loss: Number(avg.toFixed(6)) });
  console.log(`epoch ${epoch}/${EPOCHS} loss=${avg.toFixed(6)}`);
  if (avg < best) {
    best = avg;
    const json = JSON.stringify(model.toJSON());
    await writeFile(path.join(MODEL_DIR, 'matte-refiner.json'), json);
    await writeFile(path.join(CKPT_DIR, 'best.json'), json);
  }
}

const report = {
  samples: loaded.length,
  epochs: EPOCHS,
  bestLoss: best,
  model: path.join(MODEL_DIR, 'matte-refiner.json'),
  history: history.slice(-5),
};
await writeFile(path.join(REPORT_DIR, 'matte-refiner.json'), JSON.stringify(report, null, 2));
console.log(JSON.stringify(report, null, 2));

File: training/train_matte_refiner.py
#!/usr/bin/env python3
"""Train a small matte refiner; checkpoints stay under private/ (gitignored)."""

from __future__ import annotations

import json
import random
from datetime import UTC, datetime
from pathlib import Path

import numpy as np
import torch
import torch.nn as nn
from PIL import Image
from torch.utils.data import DataLoader, Dataset

ROOT = Path(__file__).resolve().parents[1]
CONFIG = json.loads((ROOT / "training" / "config.json").read_text(encoding="utf-8"))
LABELS = ROOT / CONFIG["labelsDir"]
CKPT_DIR = ROOT / CONFIG["checkpointDir"]
MODEL_DIR = ROOT / CONFIG["modelDir"]
REPORT_DIR = ROOT / CONFIG["reportDir"]
TRAIN_SIZE = int(CONFIG["trainSize"])
EPOCHS = int(CONFIG["epochs"])
BATCH = int(CONFIG["batchSize"])
LR = float(CONFIG["learningRate"])


class MatteSet(Dataset):
    def __init__(self, ids: list[str]) -> None:
        self.samples: list[tuple[Path, Path]] = []
        for sample_id in ids:
            folder = LABELS / sample_id
            if not folder.is_dir():
                continue
            original = next(folder.glob("original.*"), None)
            final = folder / "final.png"
            if original and final.is_file():
                self.samples.append((original, final))
        if not self.samples:
            raise SystemExit(f"No samples in {LABELS}. Run: node training/bootstrap-teacher.mjs")

    def __len__(self) -> int:
        return len(self.samples)

    def __getitem__(self, index: int) -> tuple[torch.Tensor, torch.Tensor]:
        original_path, final_path = self.samples[index]
        rgb = Image.open(original_path).convert("RGB")
        mask = Image.open(final_path).convert("L")
        rgb = rgb.resize((TRAIN_SIZE, TRAIN_SIZE), Image.Resampling.BILINEAR)
        mask = mask.resize((TRAIN_SIZE, TRAIN_SIZE), Image.Resampling.BILINEAR)
        x = torch.from_numpy(np.array(rgb, dtype=np.float32) / 255.0).permute(2, 0, 1)
        y = torch.from_numpy(np.array(mask, dtype=np.float32) / 255.0).unsqueeze(0)
        return x, y


class ConvBlock(nn.Module):
    def __init__(self, cin: int, cout: int) -> None:
        super().__init__()
        self.net = nn.Sequential(
            nn.Conv2d(cin, cout, 3, padding=1),
            nn.BatchNorm2d(cout),
            nn.ReLU(inplace=True),
            nn.Conv2d(cout, cout, 3, padding=1),
            nn.BatchNorm2d(cout),
            nn.ReLU(inplace=True),
        )

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return self.net(x)


class MiniMatteUNet(nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.down1 = ConvBlock(3, 32)
        self.down2 = ConvBlock(32, 64)
        self.down3 = ConvBlock(64, 128)
        self.pool = nn.MaxPool2d(2)
        self.mid = ConvBlock(128, 128)
        self.up2 = nn.ConvTranspose2d(128, 64, 2, stride=2)
        self.dec2 = ConvBlock(128, 64)
        self.up1 = nn.ConvTranspose2d(64, 32, 2, stride=2)
        self.dec1 = ConvBlock(64, 32)
        self.head = nn.Conv2d(32, 1, 1)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        d1 = self.down1(x)
        d2 = self.down2(self.pool(d1))
        d3 = self.down3(self.pool(d2))
        m = self.mid(self.pool(d3))
        u2 = self.up2(m)
        u2 = self.dec2(torch.cat([u2, d2], dim=1))
        u1 = self.up1(u2)
        u1 = self.dec1(torch.cat([u1, d1], dim=1))
        return torch.sigmoid(self.head(u1))


def dice_loss(pred: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
    smooth = 1.0
    pred = pred.reshape(pred.size(0), -1)
    target = target.reshape(target.size(0), -1)
    inter = (pred * target).sum(dim=1)
    denom = pred.sum(dim=1) + target.sum(dim=1)
    return 1 - ((2 * inter + smooth) / (denom + smooth)).mean()


def load_ids() -> list[str]:
    split = ROOT / CONFIG["privateRoot"] / "splits" / "train.txt"
    if split.is_file():
        return [line.strip() for line in split.read_text(encoding="utf-8").splitlines() if line.strip()]
    return sorted(p.name for p in LABELS.iterdir() if p.is_dir())


def main() -> None:
    CKPT_DIR.mkdir(parents=True, exist_ok=True)
    MODEL_DIR.mkdir(parents=True, exist_ok=True)
    REPORT_DIR.mkdir(parents=True, exist_ok=True)

    ids = load_ids()
    dataset = MatteSet(ids)
    loader = DataLoader(dataset, batch_size=min(BATCH, len(dataset)), shuffle=True, num_workers=0)

    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    model = MiniMatteUNet().to(device)
    opt = torch.optim.AdamW(model.parameters(), lr=LR)
    bce = nn.BCELoss()

    best = float("inf")
    history: list[dict[str, float]] = []

    for epoch in range(1, EPOCHS + 1):
        model.train()
        total = 0.0
        for x, y in loader:
            x = x.to(device)
            y = y.to(device)
            opt.zero_grad()
            pred = model(x)
            loss = bce(pred, y) + dice_loss(pred, y)
            loss.backward()
            opt.step()
            total += float(loss.item())
        avg = total / max(1, len(loader))
        history.append({"epoch": epoch, "loss": avg})
        print(f"epoch {epoch}/{EPOCHS} loss={avg:.4f} device={device}")
        torch.save({"model": model.state_dict(), "epoch": epoch, "trainSize": TRAIN_SIZE}, CKPT_DIR / "last.pt")
        if avg < best:
            best = avg
            torch.save({"model": model.state_dict(), "epoch": epoch, "trainSize": TRAIN_SIZE}, CKPT_DIR / "best.pt")

    stamp = datetime.now(UTC).strftime("%Y%m%d")
    tagged = CKPT_DIR / f"matte-refiner-{stamp}.pt"
    torch.save(torch.load(CKPT_DIR / "best.pt", weights_only=False), tagged)

    model.load_state_dict(torch.load(CKPT_DIR / "best.pt", weights_only=False)["model"])
    model.eval()
    dummy = torch.randn(1, 3, TRAIN_SIZE, TRAIN_SIZE)
    onnx_path = MODEL_DIR / "matte-refiner.onnx"
    torch.onnx.export(
        model,
        dummy,
        onnx_path,
        input_names=["rgb"],
        output_names=["alpha"],
        dynamic_axes={"rgb": {0: "batch", 2: "height", 3: "width"}, "alpha": {0: "batch", 2: "height", 3: "width"}},
        opset_version=17,
    )

    report = {
        "samples": len(dataset),
        "epochs": EPOCHS,
        "bestLoss": best,
        "checkpointDir": str(CKPT_DIR),
        "onnx": str(onnx_path),
        "history": history[-5:],
    }
    (REPORT_DIR / "matte-refiner.json").write_text(json.dumps(report, indent=2), encoding="utf-8")
    print(json.dumps(report, indent=2))


if __name__ == "__main__":
    main()
